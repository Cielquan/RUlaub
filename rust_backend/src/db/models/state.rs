use std::fmt::{self, Display, Formatter};

use diesel::prelude::*;
use diesel::result::Error;
use tracing::{debug, instrument, trace};

use super::super::{schema::states, util::last_insert_rowid};

#[derive(Queryable, Debug)]
pub struct State {
    pub id: i32,
    pub state_abbr: String,
    pub state_full: String,
}

impl State {
    pub fn new<'a>(state_abbr: &'a str, state_full: &'a str) -> NewState<'a> {
        NewState {
            state_abbr,
            state_full,
        }
    }
}

impl Display for State {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        if f.alternate() {
            write!(f, "<State {}>", self.state_full)
        } else {
            write!(
                f,
                "<State {} (Abbr: {} | ID: {})>",
                self.state_full, self.state_abbr, self.id
            )
        }
    }
}

#[derive(Insertable, Debug)]
#[table_name = "states"]
pub struct NewState<'a> {
    pub state_abbr: &'a str,
    pub state_full: &'a str,
}

impl Display for NewState<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        if f.alternate() {
            write!(f, "<State {}>", self.state_full)
        } else {
            write!(f, "<State {} (Abbr: {})>", self.state_full, self.state_abbr)
        }
    }
}

impl NewState<'_> {
    #[instrument(skip(self, conn))]
    pub fn save_to_db(self, conn: &SqliteConnection) -> Result<i32, Error> {
        debug!(target: "new_db_entry", "Adding to db: {:?}", &self);
        diesel::insert_into(states::table)
            .values(&self)
            .execute(conn)?;

        trace!(target: "new_db_entry", "Get `last_insert_rowid` for id of: {:#}", &self);
        let id = diesel::select(last_insert_rowid).get_result::<i32>(conn);
        debug!(target: "new_db_entry", "Got ID: <{:?}> for {:#}", id, &self);
        id
    }
}
