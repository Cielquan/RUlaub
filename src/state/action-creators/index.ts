export * from "./configActionCreators";
export * from "./infoPageActionCreators";
export * from "./publicHolidaysDataActionCreators";
export * from "./publicHolidaysDialogActionCreators";
export * from "./schoolHolidaysDataActionCreators";
export * from "./settingsDialogActionCreators";
export * from "./sideMenuActionCreators";
export * from "./vacationDataActionCreators";
