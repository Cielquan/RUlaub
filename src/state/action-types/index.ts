export enum CalendarRowUserMapActionType {
  UPDATE = "update calendar user order",
}

export enum ConfigActionType {
  UPDATE = "update config",
}

export enum InfoPageActionType {
  OPEN = "open InfoPage",
  CLOSE = "close InfoPage",
}

export enum PublicHolidaysDataActionType {
  ADD = "add public holidays data",
  LOAD = "load public holidays data",
  REMOVE = "remove public holidays data",
  UPDATE = "update public holidays data",
}

export enum PublicHolidaysDialogActionType {
  OPEN = "open PublicHolidays",
  CLOSE = "close PublicHolidays",
}

export enum SchoolHolidaysDataActionType {
  ADD = "add school holidays data",
  LOAD = "load school holidays data",
  REMOVE = "remove school holidays data",
  UPDATE = "update school holidays data",
}

export enum SchoolHolidaysDialogActionType {
  OPEN = "open SchoolHolidays",
  CLOSE = "close SchoolHolidays",
}

export enum SettingsDialogActionType {
  OPEN = "open SettingsDialog",
  CLOSE = "close SettingsDialog",
}

export enum SideMenuActionType {
  OPEN = "open SideMenu",
  CLOSE = "close SideMenu",
}

export enum UsersDataActionType {
  ADD = "add users data",
  LOAD = "load users data",
  REMOVE = "remove users data",
  UPDATE = "update users data",
}

export enum UsersDialogActionType {
  OPEN = "open UsersDialog",
  CLOSE = "close UsersDialog",
}

export enum VacationDialogActionType {
  OPEN = "open VacationDialog",
  CLOSE = "close VacationDialog",
}

export enum VacationTypesDataActionType {
  ADD = "add vacation types data",
  LOAD = "load vacation types data",
  REMOVE = "remove vacation types data",
  UPDATE = "update vacation types data",
}

export enum VacationTypesDialogActionType {
  OPEN = "open VacationTypesDialog",
  CLOSE = "close VacationTypesDialog",
}
