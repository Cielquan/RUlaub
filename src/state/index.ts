export * as actionCreators from "./action-creators";
export { default as store } from "./store";
export * from "./reducers";
