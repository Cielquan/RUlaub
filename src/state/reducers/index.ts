import { combineReducers } from "redux";

import dbDataReducer, { initialState as dbDataInitState } from "./dbDataReducer";
import infoPageReducer, { initialState as infoPageInitState } from "./infoPageReducer";
import languageReducer, { initialState as languageInitState } from "./languageReducer";
import localConfigReducer, {
  initialState as localConfigInitState,
} from "./localConfigReducer";
import sideMenuReducer, { initialState as sideMenuInitState } from "./sideMenuReducer";
import themeReducer, { initialState as themeInitState } from "./themeReducer";

const rootReducer = combineReducers({
  dbData: dbDataReducer,
  infoPage: infoPageReducer,
  language: languageReducer,
  localConfig: localConfigReducer,
  sideMenu: sideMenuReducer,
  theme: themeReducer,
});

export type State = ReturnType<typeof rootReducer>;

export const initialState: State = {
  dbData: dbDataInitState,
  infoPage: infoPageInitState,
  language: languageInitState,
  localConfig: localConfigInitState,
  sideMenu: sideMenuInitState,
  theme: themeInitState,
};

export default rootReducer;
