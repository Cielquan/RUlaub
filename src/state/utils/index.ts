interface DBData<T> {
  [k: string]: T;
}

export const addData = <T>(currentData: DBData<T>, dataToAdd: T[]): DBData<T> => {
  const rv: DBData<T> = JSON.parse(JSON.stringify(currentData));
  const highestID = Math.max(...Object.keys(currentData).map((id) => Number(id)));
  dataToAdd.forEach((data, idx) => {
    rv[highestID + idx + 1] = data;
  });
  return rv;
};

export const removeData = <T>(
  currentData: DBData<T>,
  IDsToRemove: string[]
): DBData<T> => {
  const rv: DBData<T> = JSON.parse(JSON.stringify(currentData));
  IDsToRemove.forEach((ID) => {
    delete rv[ID];
  });
  return rv;
};

export const updateData = <T>(
  currentData: DBData<T>,
  updatePayload: [string, T][]
): DBData<T> => {
  const rv: DBData<T> = JSON.parse(JSON.stringify(currentData));
  updatePayload.forEach((update) => {
    const [id, updatedData] = update;
    rv[id] = updatedData;
  });
  return rv;
};
