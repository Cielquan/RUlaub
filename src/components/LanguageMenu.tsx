import {
  KeyboardArrowDown as KeyboardArrowDownIcon,
  Translate as TranslateIcon,
} from "@mui/icons-material";
import {
  ListItem,
  ListItemButton,
  ListItemIcon,
  ListItemText,
  Menu,
} from "@mui/material";
import React, { MouseEvent, ReactElement, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { bindActionCreators } from "redux";

import { actionCreators, State } from "../state";
import Languages from "../state/utils/i18n";

import LanguageMenuButton from "./LanguageMenuButton";

const LanguageMenu = (): ReactElement => {
  const dispatch = useDispatch();
  const { activateDE, activateEN } = bindActionCreators(actionCreators, dispatch);
  const configState = useSelector((state: State) => state.config);
  const themeState = configState.settings.theme;
  const langState = configState.settings.language;

  const languages = [
    { changeHandle: activateDE, lang: Languages.german },
    { changeHandle: activateEN, lang: Languages.english },
  ];

  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);

  const handleClick = (event: MouseEvent<HTMLDivElement>): void => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = (): void => {
    setAnchorEl(null);
  };

  const buttonStyle = {
    color: themeState === "dark" ? "primary.main" : "action.active",
    fill: themeState === "dark" ? "primary.main" : "action.active",
  };

  return (
    <>
      <ListItem
        key="lang-btn"
        aria-controls="language-button"
        aria-haspopup="true"
        sx={{
          padding: 0,
          marginRight: 2,
        }}
      >
        <ListItemButton
          onClick={handleClick}
          sx={{
            border: "1px solid",
            borderColor: themeState === "dark" ? "primary.main" : "action.active",
            borderRadius: 2,
          }}
        >
          <ListItemIcon>
            <TranslateIcon sx={buttonStyle} />
          </ListItemIcon>
          <ListItemText
            sx={{ ...buttonStyle, minWidth: 60 }}
            primary={langState.name}
          />
          <KeyboardArrowDownIcon sx={buttonStyle} />
        </ListItemButton>
      </ListItem>
      <Menu
        data-testid="lang-menu"
        id="language-button"
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClick={handleClose}
        onClose={handleClose}
      >
        {languages.map((lang) => (
          <LanguageMenuButton
            key={lang.lang.locale}
            closeHandle={handleClose}
            changeHandle={lang.changeHandle}
            language={lang.lang}
          />
        ))}
      </Menu>
    </>
  );
};

export default LanguageMenu;
