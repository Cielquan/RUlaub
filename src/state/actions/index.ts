import { UserData } from "../../types/usersData.schema";
import {
  CalendarRowUserMapActionType,
  ConfigActionType,
  InfoPageActionType,
  PublicHolidaysDataActionType,
  PublicHolidaysDialogActionType,
  SchoolHolidaysDataActionType,
  SchoolHolidaysDialogActionType,
  SettingsDialogActionType,
  SideMenuActionType,
  UsersDataActionType,
  UsersDialogActionType,
  VacationDialogActionType,
  VacationTypesDialogActionType,
} from "../action-types";
import { ConfigPayload } from "../utils/config";
import {
  PublicHolidaysData,
  PublicHolidayDataPayload,
} from "../utils/publicHolidaysData";
import {
  SchoolHolidaysData,
  SchoolHolidayDataPayload,
} from "../utils/schoolHolidaysData";
import { UsersData, UserDataPayload } from "../utils/usersData";

export type CalendarRowUserMapAction = {
  type: CalendarRowUserMapActionType;
  payload: UsersData;
};

export type ConfigAction = {
  type: ConfigActionType;
  payload: ConfigPayload;
};

export type InfoPageAction = {
  type: InfoPageActionType;
};

export type PublicHolidaysDataLoadAction = {
  type: PublicHolidaysDataActionType.LOAD;
  payload: PublicHolidaysData;
};

export type PublicHolidaysDataUpdateAction = {
  type: PublicHolidaysDataActionType.UPDATE;
  payload: PublicHolidayDataPayload;
};

export type PublicHolidaysDataAction =
  | PublicHolidaysDataLoadAction
  | PublicHolidaysDataUpdateAction;

export type PublicHolidaysDialogAction = {
  type: PublicHolidaysDialogActionType;
};

export type SchoolHolidaysDataLoadAction = {
  type: SchoolHolidaysDataActionType.LOAD;
  payload: SchoolHolidaysData;
};

export type SchoolHolidaysDataUpdateAction = {
  type: SchoolHolidaysDataActionType.UPDATE;
  payload: SchoolHolidayDataPayload;
};

export type SchoolHolidaysDataAction =
  | SchoolHolidaysDataLoadAction
  | SchoolHolidaysDataUpdateAction;

export type SchoolHolidaysDialogAction = {
  type: SchoolHolidaysDialogActionType;
};

export type SettingsDialogAction = {
  type: SettingsDialogActionType;
};

export type SideMenuAction = {
  type: SideMenuActionType;
};

export type UsersDataAddAction = {
  type: UsersDataActionType.ADD;
  payload: UserData[];
};

export type UsersDataLoadAction = {
  type: UsersDataActionType.LOAD;
  payload: UsersData;
};

export type UsersDataRemoveAction = {
  type: UsersDataActionType.REMOVE;
  payload: string[];
};

export type UsersDataUpdateAction = {
  type: UsersDataActionType.UPDATE;
  payload: UserDataPayload[];
};

export type UsersDataAction =
  | UsersDataAddAction
  | UsersDataLoadAction
  | UsersDataRemoveAction
  | UsersDataUpdateAction;

export type UsersDialogAction = {
  type: UsersDialogActionType;
};

export type VacationDialogAction = {
  type: VacationDialogActionType;
};

export type VacationTypesDialogAction = {
  type: VacationTypesDialogActionType;
};
