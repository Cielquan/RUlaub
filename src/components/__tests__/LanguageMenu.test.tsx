import { render, screen } from "@testing-library/react";
// import userEvent from "@testing-library/user-event";
import React from "react";
import { Provider } from "react-redux";

import LanguageMenu from "../LanguageMenu";
import { initialState } from "../../state";
import generateMockStore from "../../testUtils";

describe("<LanguageMenu />", () => {
  const mockStore = generateMockStore(initialState);

  beforeEach(() => {
    mockStore.clearActions();
  });

  it("renders button correctly", () => {
    render(
      <Provider store={mockStore}>
        <LanguageMenu />
      </Provider>
    );
    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("does not render menu by default", () => {
    render(
      <Provider store={mockStore}>
        <LanguageMenu />
      </Provider>
    );
    expect(screen.queryByTestId("lang-menu")).not.toBeVisible();
  });

  // TODO:#i# fix test https://stackoverflow.com/q/69040036/16820192
  // it("renders menu correctly when opened and closed", () => {
  //   render(<LanguageMenu />);
  //   expect(screen.queryByTestId("lang-menu")).toBeNull();

  //   userEvent.click(screen.getByRole("button"));
  //   expect(screen.getByTestId("lang-menu")).toMatchSnapshot();
  //   expect(screen.getByTestId("lang-menu")).toBeVisible();

  //   // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  //   // @ts-ignore
  //   // userEvent.click(screen.getByRole("presentation").firstChild);
  //   // userEvent.click(screen.getByTestId("lang-menu"));
  //   userEvent.click(screen.getByText("Menu Button"));

  //   expect(screen.queryByTestId("lang-menu")).toBeNull();
  // });
});
