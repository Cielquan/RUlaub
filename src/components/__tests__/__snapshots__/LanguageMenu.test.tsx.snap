// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageMenu /> renders button correctly 1`] = `
<div
  class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-1rmxkrg-MuiButtonBase-root-MuiListItemButton-root"
  role="button"
  tabindex="0"
>
  <div
    class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-euno13-MuiSvgIcon-root"
      data-testid="TranslateIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m12.87 15.07-2.54-2.51.03-.03c1.74-1.94 2.98-4.17 3.71-6.53H17V4h-7V2H8v2H1v1.99h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11.76-2.04zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2l-4.5-12zm-2.62 7 1.62-4.33L19.12 17h-3.24z"
      />
    </svg>
  </div>
  <div
    class="MuiListItemText-root css-19uvsop-MuiListItemText-root"
  >
    <span
      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
    >
      English
    </span>
  </div>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-euno13-MuiSvgIcon-root"
    data-testid="KeyboardArrowDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</div>
`;
